eng/common/TestResources/TestResources-Helpers.ps1

```diff
@@ -0,0 +1,327 @@
+                }
+                else {
+                    Write-Host "Setting variable '$key': $value"
+                    $notSecretValues += $value
+                }
+                LogVsoCommand "##vso[task.setvariable variable=$key;]$value"
+            }
+            else {
+                Write-Host ($shellExportFormat -f $key, $value)
+            }
+        }
+
+        if ($key) {
+            # Isolate the environment variables for easy reading.
+            Write-Host "`n"
+            $key = $null
+        }
+    }
+
+    return $deploymentEnvironmentVariables, $deploymentOutputs
+}
+
+function SetResourceNetworkAccessRules([string]$ResourceGroupName, [array]$AllowIpRanges, [switch]$CI) {
+    SetStorageNetworkAccessRules -ResourceGroupName $ResourceGroupName -AllowIpRanges $AllowIpRanges -CI:$CI
+}
+
+function SetStorageNetworkAccessRules([string]$ResourceGroupName, [array]$AllowIpRanges, [switch]$CI, [switch]$Override) {
+    $storageAccounts = Retry { Get-AzResource -ResourceGroupName $ResourceGroupName -ResourceType "Microsoft.Storage/storageAccounts" }
+    # Add client IP to storage account when running as local user. Pipeline's have their own vnet with access
+    if ($storageAccounts) {
+        foreach ($account in $storageAccounts) {
+            $rules = Get-AzStorageAccountNetworkRuleSet -ResourceGroupName $ResourceGroupName -AccountName $account.Name
+            if ($rules -and ($Override -or $rules.DefaultAction -eq "Allow")) {
+                Write-Host "Restricting network rules in storage account '$($account.Name)' to deny access by default"
+                Retry { Update-AzStorageAccountNetworkRuleSet -ResourceGroupName $ResourceGroupName -Name $account.Name -DefaultAction Deny }
+                if ($CI -and $env:PoolSubnet) {
+                    Write-Host "Enabling access to '$($account.Name)' from pipeline subnet $($env:PoolSubnet)"
+                    Retry { Add-AzStorageAccountNetworkRule -ResourceGroupName $ResourceGroupName -Name $account.Name -VirtualNetworkResourceId $env:PoolSubnet }
+                }
+                elseif ($AllowIpRanges) {
+                    Write-Host "Enabling access to '$($account.Name)' to $($AllowIpRanges.Length) IP ranges"
+                    $ipRanges = $AllowIpRanges | ForEach-Object {
+                        @{ Action = 'allow'; IPAddressOrRange = $_ }
+                    }
+                    Retry { Update-AzStorageAccountNetworkRuleSet -ResourceGroupName $ResourceGroupName -Name $account.Name -IPRule $ipRanges | Out-Null }
+                }
+                elseif (!$CI) {
+                    Write-Host "Enabling access to '$($account.Name)' from client IP"
+                    $clientIp ??= Retry { Invoke-RestMethod -Uri 'https://icanhazip.com/' }  # cloudflare owned ip site
+                    $clientIp = $clientIp.Trim()
+                    $ipRanges = Get-AzStorageAccountNetworkRuleSet -ResourceGroupName $ResourceGroupName -Name $account.Name
```

Comment:
Everything here and below is net new. I added a check for the client IP overlapping with the AllowIpRange, in cases where the storage account has corp net IP ranges added by policy, and the script tries to add the user's IP already within said range.